카카오 로그인

rockpell opened the pull request:
# PR

### PR 종류

- [x] Code Review
- [x] New Feature
- [ ] Bug Fix
- [ ] Etc

### 내용
- 카카오 로그인 두개의 api로 스펙 변경
- https://www.notion.so/API-6144758040cc441ab3bfae78781d8c10 참조
- 로그인시 last_login 갱신을 위한 함수 추가
- 로그인시 회원가입 유무를 내려줌

rockpell committed:
[fix] user db 함수 return 포맷 수정

- 항상 rows를 포함하고 있어서 rows를 반환하도록 수정

rockpell committed:
[feat] 로그인 jwt cookie 발급

- 유저의 id를 포함한 jwt cookie를 발급

rockpell committed:
[fix] 불필요한 user 주석 제거

- user router에서 사용하지 않는 주석 제거

rockpell committed:
[feat] 에러 처리를 위한 미들웨어 추가

- router next의 인자에 값을 넣는 경우 에러로 인식하여 이 함수 호출

rockpell committed:
[fix] 카카오 로그인 스펙 변경

- 프론트에서 oauth code를 보내주는 작업을 하기 위해 스펙 변경
- login api를 이용하여 카카오 로그인을 처리하면 url에 code가 붙어옴
- 해당 code를 kakao/callback api에 보내면 서버가 처리하여 토큰을 쿠키로 내려줌

rockpell committed:
[feat] check_login 미들웨어 추가

- 현재 요청의 토큰이 유효한지 확인하는 미들웨어 추가

blingblin-g commented on backend/src/db/user.ts:
```diff
@@ -15,7 +15,7 @@ export const createUser = async (authId: string) => {
   try {
     const res = await pool.query(`SELECT * FROM USERS WHERE id = $1`, [id]);
-    return res;
+    return res.rows;
```
rows는 뭔가요?

rockpell replied:
db에서 rows에 결과 데이터를 담아서 내보내줘서 이렇게 반환하였습니다

blingblin-g commented on backend/src/middleware/check_login.ts:
```diff
@@ -0,0 +1,24 @@
+
+  const verifedToken = verifyToken(token);
+
+  console.log('verifedToken: ', verifedToken);
```
이거 지우세요!!

rockpell committed:
[fix] 불필요한 console.log 제거

blingblin-g commented on backend/src/route/user/user.ts:
```diff
@@ -1,12 +1,8 @@
 import express from 'express';
-import * as user_logic from '../../logic/user';
 import { IUserRequest } from '../../interface/interface';
```
절대경로 써주세요

rockpell replied:
넵!

blingblin-g reviewed (commented):
백엔드 로직을 잘 몰라도 이해할 수 있을 만큼 클린 코드네요!!
잘 봤습니다. 수고하셨습니다.

mimseong commented on backend/src/logic/user.ts:
```diff
@@ -17,11 +17,11 @@ export const getByAuthId = async (authId: string) => {
 export const getOrCreate = async (authId: string) => {
```
함수의 결과값 타이핑이 있으면 좋을 것 같아요
`async (authId: string): 리턴타입 =>`

mimseong replied:
async 붙은 함수는 타입을 Promise<타입> 이런 식으로 적으면 됩니다

rockpell replied:
적용해봤습니다!

mimseong commented on backend/src/middleware/check_login.ts:
```diff
@@ -0,0 +1,22 @@
+router.use(function (req, res, next) {
+  const token = req.headers.authorization;
+
+  if (!token) next('error');
```
에러 명이 조금 더 자세히 써져있으면 좋을 것 같아요

mimseong commented on backend/src/route/user/login.ts:
```diff
@@ -10,22 +10,24 @@ const router = express.Router();
 const KAKAO_CLIENT_ID = process.env.KAKAO_CLIENT_ID;
 const KAKAO_CALLBACK_URL = process.env.KAKAO_CALLBACK_URL;
 const KakaoStrategy = require('passport-kakao').Strategy;
+const cookieMaxAge: number = 1000 * 60 * 60;
```
시간 관련된 값은 dayjs 라이브러리를 쓰는게 보기에 더 좋을 것 같아요
https://www.npmjs.com/package/dayjs
한번 적용해보면 어떨까요?

rockpell replied:
이상하게 typescript에서 적용이 잘 안되네요
한번 같이 봐주시면 좋을거 같습니다

mimseong replied:
으음 확인해보니 dayjs에서 타입이 기존에 사용하려던 number랑 다른 것 같네요 일단 이렇게 두고 더 좋은 방법 찾아본 다음 적용하기로 해요!

rockpell committed:
[fix] user db, api에 리턴타입 추가

rockpell committed:
[fix] check_login 에러 내용 수정

- 에러 내용을 좀 더 상세하게 수정

rockpell committed:
[fix] user 테이블 및 인터페이스 최신 반영

- 최신 기획이 제대로 반영 안된 부분 수정

rockpell committed:
[chore] 불필요한 로그 및 함수 제거

rockpell committed:
[feat] 유저 로그인시 last_login 갱신, 회원가입 유무

- 로그인시 last_login 갱신을 위한 함수 추가
- 로그인시 회원가입 유무를 내려줌

rockpell committed:
[feat] 카카오 로그인 결과값 변경

- 대학 ID도 같이 내려주도록 변경

rockpell committed:
[fix] logout, not found의 status 변경

- not found 처리에 대한 status code 가 200으로 나가던 문제 수정
- logout api 결과가 redirect로 되어 있던 문제 수정

rockpell committed:
[fix] 로그아웃 api method 변경

- get에서 delete로 변경

rockpell committed:
[fix] 회원가입 유무를 알려주는 결과값 수정

- isSignUP이 회원가입을 한 유저라는것을 제대로 알려주지 않는것 같아 이름 변경

rockpell committed:
[fix] updateLastLogin 함수 리턴 타입

- 누락된 리턴 타입 추가

rockpell committed:
[fix] checkSignedUser 리턴 타입

- 누락된 리턴 타입 추가

SeongsangCHO commented on backend/src/db/user.ts:
```diff
@@ -12,19 +13,29 @@ export const createUser = async (authId: string) => {
-export const getUser = async (id: number) => {
+export const getUser = async (id: number): Promise<IUser> => {
   try {
     const res = await pool.query(`SELECT * FROM USERS WHERE id = $1`, [id]);
```
pool이 라이브러리인가요?

rockpell replied:
pool은 postgresql의 커넥션 풀입니다.
이걸 이용해서 커넥션을 절약 할 수 있어요

SeongsangCHO commented on backend/src/interface/interface.ts:
```diff
@@ -2,15 +2,15 @@ import { Request, Response } from 'express';
   nickname: string;
-  schoolId: number;
+  universityId: number;
   lastLogin: number;
```
lastLogin이 있는 이유는 뭔가여 ?

rockpell replied:
이거는 나중에 운영할 때를 대비해서 만든 필드인데 마지막으로 로그인한 시점이 1년이 지나거나 하면 휴면계정으로 만드려고 합니다

SeongsangCHO commented on backend/src/middleware/check_login.ts:
```diff
@@ -0,0 +1,22 @@
+
+  if (!token) next('token does not exist');
+
+  const verifedToken = verifyToken(token);
```
JWT 검증 기능은 middleware로 하면 어떤가요 ?

SeongsangCHO replied:
아 ~! 잘못봤습니당 !!
